src/blob_file_builder.cc

```diff
@@ -27,23 +128,56 @@ void BlobFileBuilder::Add(const BlobRecord& record, BlobHandle* handle) {
   if (ok()) {
     status_ = file_->Append(encoder_.GetRecord());
     num_entries_++;
-    // The keys added into blob files are in order.
-    if (smallest_key_.empty()) {
-      smallest_key_.assign(record.key.data(), record.key.size());
-    }
-    assert(cf_options_.comparator->Compare(record.key, Slice(smallest_key_)) >=
-           0);
-    assert(cf_options_.comparator->Compare(record.key, Slice(largest_key_)) >=
-           0);
-    largest_key_.assign(record.key.data(), record.key.size());
   }
 }
 
-Status BlobFileBuilder::Finish() {
+void BlobFileBuilder::WriteRawBlock(const Slice& block, BlockHandle* handle) {
+  handle->set_offset(file_->GetFileSize());
+  handle->set_size(block.size());
+  status_ = file_->Append(block);
+  if (ok()) {
+    // follow rocksdb's block based table format
+    char trailer[kBlockTrailerSize];
+    // only compression dictionary and meta index block are written
+    // by this method, we use `kNoCompression` as placeholder
+    trailer[0] = kNoCompression;
+    char* trailer_without_type = trailer + 1;
+
+    // crc32 checksum
+    auto crc = crc32c::Value(block.data(), block.size());
+    crc = crc32c::Extend(crc, trailer, 1);  // Extend to cover compression type
+    EncodeFixed32(trailer_without_type, crc32c::Mask(crc));
+    status_ = file_->Append(Slice(trailer, kBlockTrailerSize));
+  }
+}
+
+void BlobFileBuilder::WriteCompressionDictBlock(
+    MetaIndexBuilder* meta_index_builder, BlockHandle* handle) {
+  WriteRawBlock(compression_dict_->GetRawDict(), handle);
+  if (ok()) {
+    meta_index_builder->Add(kCompressionDictBlock, *handle);
+  }
+}
+
+Status BlobFileBuilder::Finish(OutContexts* out_ctx) {
   if (!ok()) return status();
 
-  std::string buffer;
+  if (builder_state_ == BuilderState::kBuffered) {
+    EnterUnbuffered(out_ctx);
+  }
+
   BlobFileFooter footer;
+  // if has compression dictionary, encode it into meta blocks
+  // and update relative fields in footer
+  if (cf_options_.blob_file_compression_options.max_dict_bytes > 0) {
+    BlockHandle meta_index_handle, uncompression_dict_handle;
```

Comment:
`uncompression_dict_handle` is useless. As commented before, no need to set `uncompression_dict_handle` to `footer`. Seems we should remove the field from `BlobFileFooter`. /cc @yiwu-arbug

Reply:
actually `uncompression_dict_handle` is removed before, and since it's useless here, I refactored `WriteCompressionDictBlock`, to allocate a block handle inside the method.